### PR TITLE
Fix corruption when a URDF file is included from a SDFormat 1.6 model

### LINKED ISSUE #2728: URDF included in SDF 1.6 is silently corrupted in Gazebo 11 when spawned from the GUI
I have one simple URDF file:
<details>
 <summary>Click to see URDF content</summary>

~~~xml
<?xml version="1.0"?>
<robot name="model_urdf">
  <link name="base_link">
    <inertial>
      <mass value="1"/>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
   <visual>
     <origin xyz="0 0 0.5" rpy="0 0 0" />
     <geometry>
       <box size="1 1 1" />
     </geometry>
     <material name="Red">
       <color rgba="1.0 0.0 0.0 1.0"/>
     </material>
   </visual>
   <collision>
     <origin xyz="0 0 0.5" rpy="0 0 0"/>
     <geometry>
       <box size="1 1 1" />
     </geometry>
   </collision>
  </link>
  <joint name="joint_a1" type="continuous">
    <origin xyz="0 0 1.0" rpy="0 -0 0"/>
    <axis xyz="0 0 -1"/>
    <parent link="base_link"/>
    <child link="link_1"/>
  </joint>
  <link name="link_1">
    <inertial>
      <mass value="1"/>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
   <visual>
     <origin xyz="0 0 0.5" rpy="0 0 0" />
     <geometry>
       <cylinder radius="0.6" length="1.0"/>
     </geometry>
     <material name="Green">
       <color rgba="0.0 1.0 0.0 1.0"/>
     </material>
   </visual>
   <collision>
     <origin xyz="0 0 0.5" rpy="0 0 0"/>
     <geometry>
       <cylinder radius="0.6" length="1.0"/>
     </geometry>
   </collision>
  </link>
</robot>
~~~

 </details>

that is included in two almost identical SDF files, one with version 1.6: 
~~~xml
<?xml version='1.0'?>
<sdf version='1.6'>
    <model name='model_sdf_1_6'>
        <include>
             <uri>model://model_urdf</uri>
             <pose>0.0 0.0 0.0 0.0 0 0.0</pose>
        </include>
    </model>
</sdf>

~~~

and another with version 1.7 : 
~~~xml
<?xml version='1.0'?>
<sdf version='1.7'>
    <model name='model_sdf_1_7'>
        <include>
             <uri>model://model_urdf</uri>
             <pose>0.0 0.0 0.0 0.0 0 0.0</pose>
        </include>
    </model>
</sdf>
~~~

when I spawn the models from the GUI, the SDF 1.7 is loaded  correctly, while the kinematics of the SDF 1.6 model are silently loaded incorrectly, see the following  GIF: 
urdf_nested_in_sdf_1_6

Interestingly, the urdf  included models is  loaded  correctly if I load it in a world model, both if the world is a SDF 1.6 or 1.7 model. 
I reproduced the problem on Gazebo 11 on both Windows and Linux. 
To  reproduce it locally, you can use this models: [urdf_included_in_sdf_1_6_mwe.zip](https://github.com/osrf/gazebo/files/4594695/urdf_included_in_sdf_1_6_mwe.zip) . 

Short setup bash script to reproduce the GUI spawning:
~~~
wget https://github.com/osrf/gazebo/files/4594695/urdf_included_in_sdf_1_6_mwe.zip
unzip urdf_included_in_sdf_1_6_mwe.zip
cd urdf_included_in_sdf_1_6_mwe
export GAZEBO_MODEL_PATH=`pwd`
gazebo
~~~

on the same terminal, you can see that loading the models from a world file is working correctly: 
~~~
gazebo world_sdf_1_6.world
gazebo world_sdf_1_7.world
~~~

### PR DESCRIPTION
**Pull request by azeey**
This fixes #2728 by sending the non-parsed text of the model to gzserver. Before this, when spawning a model using the GUI or the `gz` tool, the SDFormat/URDF file was parsed by libsdformat on the client side for error checking or previewing model placement in the world. This was also used to set a new name for the model. The parsed model was then reserialized to xml and sent to the server in a `Factory` message. With this PR, the parsing still happens for the same reasons, but instead of reserializing the model, the original text in the file is sent over a `Factory` message. To allow setting a new name, a new field ,`new_name`, is added to the `Factory` message type.

See #2728, specifically [this](https://github.com/osrf/gazebo/issues/2728#issuecomment-632401401) comment for why reserialization is a problem.

### CODE REVIEW

**Comment by traversaro** [comment]
Thanks a lot @azeey for working on this! I tried this branch on my machine, but it seems that the problem https://github.com/osrf/gazebo/issues/2728 is still present, I wounder if I have some local problem or there are still some problems after the fix in this PR.

**Comment by traversaro** [comment]
For reference, if instead I change https://github.com/osrf/gazebo/pull/2734/files#diff-b8479763685d74e12aa126230aae05ccR112 from:
~~~cxx
sdf::readFileWithoutConversion(_filename, this->dataPtr->modelSDFUnconverted,
          _filename, this->dataPtr->modelSDF, errors))
                                 errors);
~~~
to
~~~cxx
sdf::readFile(_filename, this->dataPtr->modelSDFUnconverted,
          _filename, this->dataPtr->modelSDF, errors))
                                 errors);
~~~
the behaviour is: 
gazebo_fix

so it seems that the preview is malformed, but the final spawned model is correct.

**Comment by azeey** [comment]
@traversaro Apparently I had set the sdf version to 1.7 in my `model_sdf_1_6` model for testing and forgot to change it back. As you observed, the problem is still there.

**Comment by EricCousineau-TRI** [comment]
Per VC, Addisu has some tests that should show the fix, and then this will turn into a non-draft PR.

**Comment by azeey** [comment]
I've added tests that exercise spawning a nested URDF from the GUI as well as using the `gz` tool. @traversaro, do you mind reviewing the changes I made? BTW, thanks for providing the sample files.

**Comment by traversaro** [comment]
Thanks a lot, I am checking them now.

**Review by traversaro — APPROVED** [review]
The PR seems great, especially the GUI tests. However, it would be useful if  the PR contained a small description about the fix (just a sentence, nothing to elaborate) as I far as I understand: 
> This fixes #2728 by keeping an extra model that is read in without up-conversion. This is then used for serializing and sending the model to gzserver.

Is not relevant anymore.

**Comment by azeey** [comment]
Thanks for the review @traversaro . I've updated the PR description.

File: gazebo/msgs/factory.proto [hunk 1]
```diff
@@ -40,4 +40,7 @@ message Factory
   /// \brief Whether the server is allowed to rename the model in case of
   /// overlap with existing models.
   optional bool allow_renaming = 6 [default = true];
+
+  /// \brief New name of the entity
+  optional string new_name = 7;
```
Comment (scpeters) [inline gazebo/msgs/factory.proto]: modifying protobuf messages usually breaks ABI

our ABI checking CI jobs have been broken for a while, though I think they were recently fixed; I'll try to run one for this branch

Comment (scpeters) [reply gazebo/msgs/factory.proto]: [Build Status](https://build.osrfoundation.org/job/gazebo-abichecker-any_to_any-ubuntu_auto-amd64/194/) https://build.osrfoundation.org/job/gazebo-abichecker-any_to_any-ubuntu_auto-amd64/194/

Comment (scpeters) [reply gazebo/msgs/factory.proto]: yeah, the ABI checker is not happy about this:

* https://build.osrfoundation.org/job/gazebo-abichecker-any_to_any-ubuntu_auto-amd64/194/API_5fABI_20report/

Comment (azeey) [reply gazebo/msgs/factory.proto]: Ouch!. I guess we'll have to modify the xml to set the new name of the model. That will complicate things, but I'll give it a shot.

Comment (azeey) [reply gazebo/msgs/factory.proto]: I removed the field in https://github.com/osrf/gazebo/pull/2734/commits/5a58db6d4730fcc64bac662165cd688291aeedae. The solution I came up with is to use tinyxml to set the new name of the model to be spawned.

File: tools/gz.cc [hunk 2]
```diff
@@ -482,7 +483,26 @@ bool ModelCommand::RunImpl()
...
+    std::string resolvedFileName = filename;
+    if (common::isDirectory(filename))
+    {
+      // Use sdf::getModelFilePath() instead of filename because filename
+      // might be a model rectory
+      resolvedFileName = sdf::getModelFilePath(filename);
+      if (resolvedFileName.empty())
+      {
+        std::cerr
+            << "The provided file is a directory, but a model could not found"
```
Comment (scpeters) [inline tools/gz.cc]: > could not found

could not *be* found

File: tools/gz.cc [hunk 3]
```diff
@@ -549,27 +569,54 @@ bool ModelCommand::RunImpl()
...
     transport::NodePtr _node)
 {
-  sdf::ElementPtr modelElem = _sdf->Root()->GetElement("model");
+  transport::PublisherPtr pub = _node->Advertise<msgs::Factory>("~/factory");
+  pub->WaitForConnection();
+
+  // To set the name of the model, we have to parse the file, set the name and
+  // reserialize to SDFormat. However, loading it using libsdformat will result
+  // in upconversion, which we would like to avoid because the resulting
+  // document will lose its OriginalVersion attribute.
```
Comment (scpeters) [inline tools/gz.cc]: maybe add one more bit: "So instead we parse and edit the model name with TinyXML"

File: gazebo/common/SdfFrameSemantics.cc [hunk 4]
```diff
@@ -75,15 +83,18 @@ void convertPosesToSdf16(const sdf::ElementPtr &_modelElem)
...
   sdf::Errors errors = modelSDFDom.Load(_modelElem);
-  if (!errors.empty())
+
+  for (const auto &error : errors)
   {
-    for (const auto &error : errors)
+    if (isSdfFrameSemanticsError(error))
     {
       gzerr << error << "\n";
     }
```
Comment (j-rivero) [inline gazebo/common/SdfFrameSemantics.cc:92]: is it safe to ignore the errors that are not FrameSemanticsErrors or they are handled in a different place?

File: gazebo/gui/ModelMaker.cc [hunk 5]
```diff
@@ -116,6 +118,14 @@ bool ModelMaker::InitFromFile(const std::string &_filename)
     }
     return false;
   }
+
+  // Since the above didn't fail, we assume the file exists
+  // We also assume that _filename is a path to a file, not to a model directory
```
Comment (j-rivero) [inline gazebo/gui/ModelMaker.cc:123]: is this assumption safe enough or should we add a check for a file?

File: gazebo/common/SdfFrameSemantics.cc (hunk 4 above)
Comment (j-rivero) [reply gazebo/common/SdfFrameSemantics.cc:92]: I see that it is a common pattern in more files under this PR.

**Review by j-rivero — COMMENTED** [review]
Quick look at the code and added some questions.

File: gazebo/common/SdfFrameSemantics.cc (hunk 4 above)
Comment (azeey) [reply gazebo/common/SdfFrameSemantics.cc:92]: DOM objects are only used for frame semantics in Gazebo classic. We use the `sdf::Element` API for actually loading SDFormat worlds and models. So, IMO, errors reported by DOM objects are only relevant if they are related to frame semantics.

File: gazebo/gui/ModelMaker.cc (hunk 5 above)
Comment (azeey) [reply gazebo/gui/ModelMaker.cc:123]: As far as I can tell, all the code paths that call `InitFromFile` have a file as an argument. 
* From tests in https://github.com/osrf/gazebo/blob/gazebo11/gazebo/gui/ModelMaker_TEST.cc
* From InsertModelWidget via a `createEntity` event https://github.com/osrf/gazebo/blob/7200818c3f1b905505d25c66d5ad004cd0003e88/gazebo/gui/InsertModelWidget.cc#L319-L326

File: tools/gz.cc (hunk 2 above)
Comment (azeey) [reply tools/gz.cc]: https://github.com/osrf/gazebo/pull/2734/commits/60d082bcdde5f998286b46c7957ca14a2988b1cb

File: tools/gz.cc (hunk 3 above)
Comment (azeey) [reply tools/gz.cc]: https://github.com/osrf/gazebo/pull/2734/commits/60d082bcdde5f998286b46c7957ca14a2988b1cb

**Review by scpeters — APPROVED** [review]

**Pull request merged**